### PR TITLE
fix(tyck): fix missing simple_cast

### PR DESCRIPTION
**Pull request by andylokandy**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

Closes #issue

### CODE REVIEW

**Comment by RinChanNOWWW** [comment]
Here are some unit tests:

```rust
fn test_cast_to_nested_type(file: &mut impl Write, is_try: bool) {
    let prefix = if is_try { "TRY_" } else { "" };

    run_ast(
        file,
        format!("{prefix}CAST(((1, TRUE), 1) AS Tuple(Tuple(INT, INT), INT))"),
        &[],
    );
    run_ast(
        file,
        format!("{prefix}CAST(((1, 'a'), 1) AS Tuple(Tuple(INT, INT NULL), INT))"),
        &[],
    );
    run_ast(
        file,
        format!("{prefix}CAST(((1, 'a'), 1) AS Tuple(Tuple(INT, INT), INT) NULL)"),
        &[],
    );
    run_ast(
        file,
        format!("{prefix}CAST([(1,TRUE),(2,FALSE)] AS Array(Tuple(INT, INT)))"),
        &[],
    );
    run_ast(
        file,
        format!("{prefix}CAST([(1,'a'),(2,'a')] AS Array(Tuple(INT, INT)) NULL)"),
        &[],
    );
    run_ast(
        file,
        format!("{prefix}CAST([(1,'a'),(2,'a')] AS Array(Tuple(INT, INT NULL)))"),
        &[],
    );
    run_ast(
        file,
        format!("{prefix}CAST([[TRUE], [FALSE, TRUE]] AS Array(Array(INT)))"),
        &[],
    );
    run_ast(
        file,
        format!("{prefix}CAST([['a'], ['b', 'c']] AS Array(Array(INT) NULL))"),
        &[],
    );
}

```

File: src/query/functions/tests/it/scalars/testdata/cast.txt [hunk 1]
```diff
@@ -3137,3 +3204,75 @@ evaluation (internal):
...
+output         : ((1_i32, NULL), 1_i32)
+
+
+ast            : TRY_CAST(((1, 'a'), 1) AS Tuple(Tuple(INT, INT), INT) NULL)
+raw expr       : TRY_CAST(tuple(tuple(1_u8, "a"), 1_u8) AS ((Int32, Int32), Int32) NULL)
+checked expr   : TRY_CAST(tuple<(UInt8, String), UInt8>(tuple<UInt8, String>(1_u8, "a"), 1_u8) AS ((Int32 NULL, Int32 NULL) NULL, Int32 NULL) NULL)
+optimized expr : ((1_i32, NULL), 1_i32)
+output type    : ((Int32 NULL, Int32 NULL) NULL, Int32 NULL) NULL
+output domain  : (({1..=1}, {NULL}), {1..=1})
+output         : ((1_i32, NULL), 1_i32)
```
Comment (RinChanNOWWW) [inline src/query/functions/tests/it/scalars/testdata/cast.txt:3231]: If this should output (NULL) ?

File: src/query/functions/tests/it/scalars/testdata/cast.txt [hunk 2]
```diff
@@ -3137,3 +3204,75 @@ evaluation (internal):
...
+output         : [(1_i32, 1_i32), (2_i32, 0_i32)]
+
+
+ast            : TRY_CAST([(1,'a'),(2,'a')] AS Array(Tuple(INT, INT)) NULL)
+raw expr       : TRY_CAST(array(tuple(1_u8, "a"), tuple(2_u8, "a")) AS Array((Int32, Int32)) NULL)
+checked expr   : TRY_CAST(array<T0=(UInt8, String)><T0, T0>(tuple<UInt8, String>(1_u8, "a"), tuple<UInt8, String>(2_u8, "a")) AS Array((Int32 NULL, Int32 NULL) NULL) NULL)
+optimized expr : [(1_i32, NULL), (2_i32, NULL)]
+output type    : Array((Int32 NULL, Int32 NULL) NULL) NULL
+output domain  : [({1..=2}, {0..=0} ∪ {NULL})]
+output         : [(1_i32, NULL), (2_i32, NULL)]
```
Comment (RinChanNOWWW) [inline src/query/functions/tests/it/scalars/testdata/cast.txt:3249]: If this should output [NULL, NULL]?

File: src/query/functions/tests/it/scalars/testdata/cast.txt [hunk 3]
```diff
@@ -3137,3 +3204,75 @@ evaluation (internal):
...
+output         : [[1_i32], [0_i32, 1_i32]]
+
+
+ast            : TRY_CAST([['a'], ['b', 'c']] AS Array(Array(INT) NULL))
+raw expr       : TRY_CAST(array(array("a"), array("b", "c")) AS Array(Array(Int32) NULL))
+checked expr   : TRY_CAST(array<T0=Array(String)><T0, T0>(array<T0=String><T0>("a"), array<T0=String><T0, T0>("b", "c")) AS Array(Array(Int32 NULL) NULL) NULL)
+optimized expr : [[NULL], [NULL, NULL]]
+output type    : Array(Array(Int32 NULL) NULL) NULL
+output domain  : [[{0..=0} ∪ {NULL}]]
+output         : [[NULL], [NULL, NULL]]
```
Comment (RinChanNOWWW) [inline src/query/functions/tests/it/scalars/testdata/cast.txt:3276]: If this should output [NULL, NULL]?

File: src/query/functions/tests/it/scalars/testdata/cast.txt (hunk 1 above)
Comment (andylokandy) [reply src/query/functions/tests/it/scalars/testdata/cast.txt:3231]: @sundy-li PTAL

Comment (sundy-li) [reply src/query/functions/tests/it/scalars/testdata/cast.txt]: I think the output type should be `((Int32 NULL, Int32 NULL) NULL, Int32 NULL)`, we should forbidden the use `Tuple` or `Array` inside in `Nullable`.

Comment (sundy-li) [reply src/query/functions/tests/it/scalars/testdata/cast.txt]: If the original type is not matched, so all the nested types should be null.

Comment (andylokandy) [reply src/query/functions/tests/it/scalars/testdata/cast.txt]: `TRY_CAST(1 AS TUPLE(INT, INT))` has to output `Null`, therefore the output type of `TRY_CATS(xxx AS TUPLE(INT, INT))` is `TUPLE(Int32 NULL, Int32 NULL) NULL`, which has Tuple inside nullable.

Comment (RinChanNOWWW) [reply src/query/functions/tests/it/scalars/testdata/cast.txt]: Nullable nested types are unavoidable in runtime.

Comment (sundy-li) [reply src/query/functions/tests/it/scalars/testdata/cast.txt]: Ok, so we should allow this in runtime evaluation.

Comment (sundy-li) [reply src/query/functions/tests/it/scalars/testdata/cast.txt]: > Nullable nested types are unavoidable in runtime.

It's forbidden in DDL but can be available in runtime.

Comment (andylokandy) [reply src/query/functions/tests/it/scalars/testdata/cast.txt]: There are two types of errors CAST could throw:

1. trying to cast to unsupported type.
2. The cast overflowed.

So there is another solution to this problem: Allowing `TRY_CAST` to throw errors on the first error type.

Comment (sundy-li) [reply src/query/functions/tests/it/scalars/testdata/cast.txt]: >  Allowing TRY_CAST to throw errors on the first error type.
That could be better.

**Review by sundy-li — APPROVED** [review]

**Review by RinChanNOWWW — APPROVED** [review]

**Pull request merged**